warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java

```diff
@@ -0,0 +1,235 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static datawave.query.jexl.JexlASTHelper.nodeToKey;
+
+/**
+ * This class generates the union of two {@link IndexInfo} objects. It handles all combinations of document and infinite ranges and will add in any nodes that
+ * were delayed.
+ * <p>
+ * Consider the following example, where the query 'A OR B' is executed against the index. Suppose that term A hits on documents 1 and 2, and term b hits on
+ * documents 2 and 3. Logically, we can represent this as,
+ * <p>
+ * <code>
+ * A = {doc1, doc2}
+ * B = {doc2, doc3}
+ * </code>
+ * <p>
+ * This is represented in the code as IndexInfo objects with an IndexMatch object for each hit in the index.
+ * <p>
+ * <code>
+ * IndexInfo(A)
+ * doc1 = {A}
+ * doc2 = {A}
+ * 
+ * IndexInfo(B)
+ * doc2 = {B}
+ * doc3 = {B}
+ * </code>
+ * <p>
+ * The union of IndexInfo(A) and IndexInfo(B) produce the following,
+ * <p>
+ * <code>
+ * IndexInfo(A OR B)
+ * doc 1 = {A}
+ * doc 2 = {A, B}
+ * doc 3 = {B}
+ * </code>
+ */
+public class IndexInfoUnion {
+    
+    public static IndexInfo union(IndexInfo left, IndexInfo right) {
+        return union(left, right, new ArrayList<>());
+    }
+    
+    public static IndexInfo union(IndexInfo left, IndexInfo right, List<JexlNode> delayedNodes) {
+        
+        // The merged node list is constructed independently of the IndexMatch collection.
+        // We do this because at this point we don't know if we are building document ranges or shard/day ranges.
+        // See TupleToRange for how the IndexInfo object is used to build query ranges.
+        Map<String,JexlNode> mergedNodes = new HashMap<>();
+        
+        // 1. Unwrap the top level Jexl node for the IndexInfo objects and merge them together.
+        mergedNodes = addJexlNodesToMap(mergedNodes, left.myNode);
+        mergedNodes = addJexlNodesToMap(mergedNodes, right.myNode);
+        
+        // 2. Add delayed nodes to the merged nodes
+        mergedNodes = addJexlNodesToMap(mergedNodes, delayedNodes);
+        
+        // 3. Generate the inverted multimap of uids to jexl nodes.
+        Set<IndexMatch> matches = mergeIndexMatches(left.uids, right.uids, delayedNodes);
+        
+        // 4. Build the final IndexInfo object.
+        IndexInfo merged = new IndexInfo();
+        
+        // 5. Handle the uids
+        if (left.isInfinite() || right.isInfinite()) {
+            // If either side is infinite we cannot perform a document uid specific lookup.
+            merged.uids = ImmutableSortedSet.of();
+        } else {
+            merged.uids = ImmutableSortedSet.copyOf(matches);
+        }
+        
+        // 6. Handle uid count
+        if (left.isInfinite() || right.isInfinite()) {
+            merged.count = -1;
+        } else {
+            merged.count = merged.uids.size();
+        }
+        
```

Comment:
Could these be combined since the condition is the same, or is there something else I am missing?

Reply:
They could be combined. I'll do that while resolving some rebase conflicts.